store/src/java/com/zimbra/cs/mailbox/Mailbox.java

```diff
@@ -9064,6 +9071,63 @@ public Mountpoint refreshMountpoint(OperationContext octxt, int mountpointId, St
         return getMountpointById(octxt, mountpointId);
     }
 
+    public void createFileSharedWithMe(Mailbox mbox, OperationContext octxt, int folderId, String fileName,
+            String ownerAccountId, int ownerFileId, String remoteUuid, String fileOwnerName,
+            String contentType, long fileSize, String rights) throws ServiceException {
+
+        CreateFileSharedWithMe redoRecorder = new CreateFileSharedWithMe(mId, folderId, fileName, ownerAccountId,
+                ownerFileId, remoteUuid, fileOwnerName, contentType, rights);
+
+        boolean success = false;
+        try {
+            beginTransaction("createFileSharedWithMe", octxt, redoRecorder);
+            CreateFileSharedWithMe redoPlayer = (CreateFileSharedWithMe) currentChange().getRedoPlayer();
+
+            int mptId = getNextItemId(redoPlayer == null ? ID_AUTO_INCREMENT : redoPlayer.getId());
+            String uuid = redoPlayer == null ? UUIDUtil.generateUUID() : redoPlayer.getUuid();
+            UnderlyingData data = new UnderlyingData();
+            data.uuid = remoteUuid + ":" + ownerFileId;
+            data.id = mptId;
+            data.type = Type.DOCUMENT.toByte();
+            data.folderId = folderId;
+            data.date = mbox.getOperationTimestamp();
+            data.name = fileName;
+            data.setSubject(fileName);
+            data.size = fileSize;
+            data.metadata = encodeFileSharedMetadata(new Metadata(), fileOwnerName, ownerAccountId,
+                    ownerFileId, contentType, rights).toString();
+            data.contentChanged(mbox);
+            new DbMailItem(mbox).create(data);
+            success = true;
+        } finally {
+            endTransaction(success);
+        }
+    }
+
+    public void deleteFileSharedWithMe(Mailbox mbox, OperationContext octxt, String remoteUuid)
+            throws ServiceException {
+        CreateFileSharedWithMe redoRecorder = new CreateFileSharedWithMe(mId, remoteUuid);
+        boolean success = false;
+        try {
+            beginTransaction("deleteFileSharedWithMe", octxt, redoRecorder);
+            CreateFileSharedWithMe redoPlayer = (CreateFileSharedWithMe) currentChange().getRedoPlayer();
```

Comment:
* Is it okay to use **Create**FileSharedWithMe here? Don't we need to to define DeleteFileSharedWithMe or like that?
* this `redoPlayer` is not used.

Reply:
Please verify that playing redolog can restore the change(s) correctly at testing phase as well.

Reply:
@k-kato don't have much idea about this API but you meant when messages are permanently deleted and user recovers deleted mails, this emails should be recovered and shared links should work?

Reply:
@jiteshsojitra In `createFileSharedWithMe`, `deleteFileSharedWithMe` and `updateFileSharedWithMe` functions in Mailbox.java, DB entry is created/updated/deleted for an item which is shared with the account.
For example, after full back up is taken, a file is shared (or the sharing is updated or revoked). The process is recorded in redolog. When an account is restored from backup data and redolog, the target record should be restored (= created/updated/deleted) by replaying the redolog.
We need to check all cases; create, update and delete.

Reply:
marking this as resolved as this needs to be in testing phase